### ISSUE TITLE
UI-Select value field

### ISSUE THREAD

**Issue by gendalf** [JosephusPaye/Keen-UI#107]
Please, add value field in options for write to (:value). Its so unnaturally every time making temporary variable, lookup value from options to this variable, and then watch for changes for get new value only.

example
Options for list = [{ text: 'variant1', value:'id1' }, { text: 'variant2', value:'id2' }]
data model = { .... 'selectvalue': 'id1' }

**Comment by EmilMoe** [comment]
Already when you change the selected item the value property changes. I'm not sure I understand your question.

**Comment by gendalf** [comment]
From KeenUI select docs:
"The entire option object is written to the value prop when the user makes a selection."

But in 90% no reason store in data model all option (with label in few languages, icons etc..), only ID value need. Some "options-value-field" option in select component.
Example
options_list = [{ text: 'variant1', value:'id1' }, { text: 'variant2', value:'id2' }]
ui-select(:options="options_list", :value.sync="model.optionA", options-value-field="value")

When set model.optionA = "id1" its select option 1. When select from list it write options_list[options-value-field] to :value, not whole option object.

**Comment by EmilMoe** [comment]
So when you want to pull the value with AJAX you want to do it this way instead, correcy?

~~~~javascript
this.list[$refs.select.value].id
~~~~

Instead of this

~~~~javascript
this.$refs.select.value.id;
~~~~

I don't agree that it's more natural to only store the ID as value and not a whole object. The only argument would be performance, but I don't think this would optimize performance significantally.

**Comment by gendalf** [comment]
Default html select component mode:
<select :model="record.country">
   <option v-for="country in countries" value="{{country._id}}">{{country.name}}</option>
   // <option value="US">USA</option> <option value="CA">Canada</option>
   .....
</selected>
I have no code, just put model and render list on options...
For example 1 million records, and you don't agree to store only id? Its max simple example, option can be in few languages, have additional description fields etc... If I will change some letters in options or add field with flag image path, I will update all database?

UI Select is good work! But to use it I need....
<ui-select :value.sync="temp_country" :options="countries">
1. Define temporary variable for value
2. Lookup value and fill temp variable temp_country = _.findBy(countries, { _id: record.country })
3. Watch for changes on temp_country, and than apply changes to record.country = temp_country._id
Why I need temp_country ? Because ui-select is rewrite my record.country to null by default, if it cant lookup it in options... this also bad idea ... change model without user action... on component init..

Standard select, selectize, select2, chosen all allow standard approach with id.  I think both methods must be supported. 

Also I thinks will be very helpful to allow use slot with options, not only bind variable. For example for bind enable/disable for some options or for master-slave selection structures in template.

**Comment by EmilMoe** [comment]
You might be right, I'm not sure I get everything correctly. I would still see value as a pointer to an index in options, but when calling value it should return that part of options so the behaviour would feel the same. But that would definitely give the option also to apply an ID to value, on the other hand you would still need to know the exact index so why not just let that be done behind the scenes?

If you have 1 mio entries, I would highly recommend not to show that on a list at the same time. Use the search and chunk load feature.

**Comment by gendalf** [comment]
1 m entries - I mean records in database, not options values.

behind the scenes?
I has write before what manipulation with variable I must todo to make ui-select connect with model. With temp variables, additional watchers and dirty code :\ - its wrong every time duplicate this base manipulation, its must be inside component feature.
Or you mean behind the scenes on server?
I can not imagine how: fetch records from db -> fetch options list -> convert  id to option object -> send to client side?  and than reverse actions on save? no way :)

my suggestion:
If pass param like "lookup-key", ui-select automatic use key for lookup in options, and store to model only options[lookup-key]
`<ui-select :value.sync="record.country" :options="countries" lookup-key="id">`
or if lookup-key is not set, ui-select manipulate with object it is now implemented.

**Comment by EmilMoe** [comment]
So what you want is an attribute that defines which key to search on? I guess that makes sense.

I would be something like this

~~~~javascript
if (this.lookupKey)
  for (var i = 0; i < this.options.length; i++)
    if (this.value[this.lookupKey] == this.options[this.lookupKey])
      return this.options[this.lookupKey];
~~~~

 Anyway I seem to have troubles creating default selections, not sure if it's wrong to return a whole object?

**Comment by gendalf** [comment]
```
if (this.lookupKey)
  for (var i = 0; i < this.options.length; i++)
    if (this.value == this.options[this.lookupKey])
      return this.options[this.lookupKey];
```
If lookupKey exists value is key, not object. But ideal variant is
```
if (this.lookupKey)
  for (var i = 0; i < this.options.length; i++)
    if (this.value == this.options[this.lookupKey] || (this.value && this.value[this.lookupKey] == this.options[this.lookupKey]))
      return this.options[this.lookupKey];
```
Now will working key value, and value { id: 'US', name:'USA' } == option { name:'USA', id: 'US' }

About default value, I thinks is just that developer must determine by hand. It wrong way to rewrite model if options not loaded yet via ajax or not exists, selection must be empty like do it default html select component.

**Comment by JosephusPaye** [comment]
For setting the selected option programmatically from outside the component, it will be possible in the next release, using a new `ui-select::set-selected` event.

As for writing only the option's `.value` property to the `:value` prop when a selection is made, we can't do that because we use the object to highlight the currently selected value.

For the lookup key, the next version allows for customizing the `text` and `value` keys to fit your data.

**Comment by JosephusPaye** [comment]
Everything described in my [comment above](https://github.com/JosephusPaye/Keen-UI/issues/107#issuecomment-238015030) now available in [v0.8.9](https://github.com/JosephusPaye/Keen-UI/releases/tag/v0.8.9). See the [UiSelect docs](https://josephuspaye.github.io/Keen-UI/#/ui-select-docs).
